digester/parser.go

```diff
@@ -0,0 +1,310 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/enescakir/emoji"
+	"github.com/gomarkdown/markdown"
+	"github.com/gomarkdown/markdown/html"
+	"github.com/gomarkdown/markdown/parser"
+)
+
+type Block struct {
+	Type     string    `json:"type"`
+	Elements []Element `json:"elements"`
+}
+
+type Element struct {
+	Type           string    `json:"type"`
+	Elements       []Element `json:"elements"`
+	Text           string    `json:"text"`
+	Style          Style     `json:"style"`
+	Indent         int       `json:"indent"`
+	Border         int       `json:"border"`
+	EmojiName      string    `json:"name"`
+	URL            string    `json:"url"`
+	UserID         string    `json:"user_id"`
+	ChannelID      string    `json:"channel_id"`
+	BroadcastRange string    `json:"range"`
+	ColorValue     string    `json:"value"`
+}
+
+type Style struct {
+	IsList    bool
+	ListStyle string
+	TextStyle TextStyle
+}
+
+type TextStyle struct {
+	Bold   bool `json:"bold"`
+	Italic bool `json:"italic"`
+	Strike bool `json:"strike"`
+	Code   bool `json:"code"`
+}
+
+func (s *Style) UnmarshalJSON(data []byte) error {
+	var listStyle string
+	if err := json.Unmarshal(data, &listStyle); err == nil {
+		s.IsList = true
+		s.ListStyle = listStyle
+		return nil
+	}
+
+	var textStyle TextStyle
+	if err := json.Unmarshal(data, &textStyle); err == nil {
+		s.IsList = false
+		s.TextStyle = textStyle
+		return nil
+	}
+
+	return fmt.Errorf("unknown style type")
+}
+
+func parseText(element Element) string {
+	result := ""
+
+	if element.Type != "text" {
+		fmt.Println("[WARNING] Element is not text")
+		return result
+	}
+
+	if element.Style.IsList {
+		fmt.Println("[WARNING] List element tried to be parsed as text")
+		return result
+	}
+
+	leadingSpacesCount := 0
+	for _, char := range element.Text {
+		if char == ' ' {
+			leadingSpacesCount++
+		} else {
+			break
+		}
+	}
+
+	if leadingSpacesCount == len(element.Text) {
+		return element.Text
+	}
+
+	trailingSpacesCount := 0
+	for i := len(element.Text) - 1; i >= 0; i-- {
+		if element.Text[i] == ' ' {
+			trailingSpacesCount++
+		} else {
+			break
+		}
+	}
+
+	result = element.Text[leadingSpacesCount : len(element.Text)-trailingSpacesCount]
+
+	if element.Style.TextStyle.Bold {
+		result = "**" + result + "**"
+	}
+	if element.Style.TextStyle.Italic {
+		result = "*" + result + "*"
+	}
+	if element.Style.TextStyle.Strike {
+		result = "~~" + result + "~~"
+	}
+	if element.Style.TextStyle.Code {
+		result = "`" + result + "`"
+	}
```

Comment:
would be nice to have this in a function, again more elegant

Reply:
did you mean Switch by any chance?

Reply:
umm no. meant a function itself here. though becoming a boilerplate and quite redundant code, having simple functions to deal such stuff helps, though only if it is being reused elsewhere. If this is a one time thing, feel free to ignore the comment and resolve it.